File: src/main/java/mffs/tile/ManipulatorCalculationThread.java
package mffs.tile;

import java.util.HashSet;
import java.util.Set;

import universalelectricity.api.vector.Vector3;

/**
 * A thread that allows multi-threading calculation of projector fields.
 * 
 * @author Calclavia
 * 
 */
public class ManipulatorCalculationThread extends Thread
{
	public interface IThreadCallBack
	{
		/**
		 * Called when the thread finishes the calculation.
		 */
		public void onThreadComplete();
	}

	private TileForceManipulator manipulator;
	private IThreadCallBack callBack;

	public ManipulatorCalculationThread(TileForceManipulator projector)
	{
		this.manipulator = projector;
	}

	public ManipulatorCalculationThread(TileForceManipulator projector, IThreadCallBack callBack)
	{
		this(projector);
		this.callBack = callBack;
	}

	@Override
	public void run()
	{
		this.manipulator.isCalculatingManipulation = true;

		try
		{
			/**
			 * Move
			 */
			Set<Vector3> mobilizationPoints = this.manipulator.getInteriorPoints();

			if (this.manipulator.canMove())
			{
				this.manipulator.manipulationVectors = new HashSet<Vector3>();

				for (Vector3 position : mobilizationPoints)
				{
					this.manipulator.manipulationVectors.add(position.clone());
				}
			}
			else
			{
				this.manipulator.markFailMove = true;
			}
		}
		catch (Exception e)
		{
			e.printStackTrace();
		}

		this.manipulator.isCalculatingManipulation = false;

		if (this.callBack != null)
		{
			this.callBack.onThreadComplete();
		}

	}
}

File: src/main/scala/mffs/tile/ManipulatorCalculationThread.scala
package mffs.tile

import java.util.HashSet
import universalelectricity.api.vector.Vector3
import calclavia.api.mffs.fortron.IServerThread
import scala.collection.JavaConversions._

/**
 * A thread that allows multi-threading calculation of projector fields.
 *
 * @author Calclavia
 *
 */
class ManipulatorCalculationThread(manipulator: TileForceManipulator) extends Thread with IServerThread
{
	private var callBack: IThreadCallBack = null

	def this(manipulator: TileForceManipulator, callBack: IThreadCallBack)
	{
		this(manipulator)
		this.callBack = callBack
	}

	override def run
	{
		manipulator.isCalculatingManipulation = true

		try
		{
			val mobilizationPoints = this.manipulator.getInteriorPoints

			if (manipulator.canMove())
			{
				manipulator.manipulationVectors = new HashSet[Vector3]
				mobilizationPoints.foreach(pos => manipulator.manipulationVectors.add(pos.clone()))
			}
			else
			{
				manipulator.markFailMove = true
			}
		}
		catch
			{
				case e: Exception =>
				{
					e.printStackTrace
				}
			}

		manipulator.isCalculatingManipulation = false

		if (callBack != null)
		{
			callBack.onThreadComplete
		}
	}

}